File: sim.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import scipy
from typing import Tuple
from tqdm import trange

# Physical parameters

b = 4 # Ns/m
L = .25 # m
c = 200 # m/s

# Simulation parameters

n = 200
delta_x = L/n # m (spacial resolution)
delta_t = delta_x / (c * np.sqrt(2)) # minimum delta_t for numerical stability
f = 1/delta_t
print('Sampling frequency is', f, 'hz')

# --- Initial state ---

def pluck(l: float, delta_y: float) -> np.ndarray:
    '''Get y with an initial state, plucking the string at length l with displacement delta_y'''
    y = np.zeros(n)
    n_displaced = int(n*(l/L))
    for i in range(1, n_displaced):
        y[i] = i/n_displaced * delta_y
    
    for i in range(n_displaced, n-1):
        y[i] = (1-(i-n_displaced+1)/(n-n_displaced)) * delta_y

    return y

def gaussian_smooth(arr: np.ndarray, sigma: float) -> np.ndarray:
    '''Smooths a 1D NumPy array using a Gaussian kernel.'''
    kernel = np.exp(-((np.arange(len(arr)) - len(arr) // 2) ** 2) / (2 * sigma ** 2))
    kernel /= np.sum(kernel)
    smoothed_arr = np.convolve(arr, kernel, mode='same')
    return smoothed_arr

# --- derivation ---

def partial_x2(y: np.ndarray) -> np.ndarray:
    '''Second spatial derivative using central differences'''
    d2y_dx2 = (np.roll(y, -1) - 2 * y + np.roll(y, 1)) / delta_x**2
    d2y_dx2[0] = 0  # Dirichlet boundary conditions
    d2y_dx2[-1] = 0
    return d2y_dx2

# --- simulation ---

def timestep(y: np.ndarray, v: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
    '''Performs one time step of the wave equation'''
    d2y_dx2 = partial_x2(y)
    d2y_dt2 = c**2 * d2y_dx2 - b * v
    v += d2y_dt2 * delta_t
    y += v       * delta_t
    return y, v

# --- recording

def normalize(a: np.ndarray) -> np.ndarray:
        a -= np.min(a)
        a /= np.max(a)
        return (a - .5) * 2

def perform_recording(y, length: float, microphone_pos: float, filename: str) -> None:
    v = np.zeros(n)
    n_recording = int(length / delta_t)
    microphone = np.zeros(n_recording)

    for i in trange(n_recording):
        y, v = timestep(y, v)
        microphone[i] = (y[int(n * microphone_pos/L)])

    audio = normalize(microphone)
    scipy.io.wavfile.write(filename=filename, rate=int(f), data=audio)

# --- visualization ---

def visualize(y):
    x = np.linspace(0, L, n)
    v = np.zeros(n)

    fig, ax = plt.subplots()
    line, = ax.plot(x, y)
    ax.set_ylim(-max(y), max(y))  # Set the y-axis limits to fit the initial condition

    def animate(i):
        global y
        global v
        y, v = timestep(y, v)
        line.set_ydata(y)
        return line,

    ani = animation.FuncAnimation(fig, animate, frames=10, interval=1, blit=True)
    plt.show()

if __name__ == '__main__':
    y_0 = pluck(0.1, 0.007)
    visualize(y_0)

File: sim.js
// Constants and initial conditions
const b = 4; // Ns/m
const L = 0.25; // m
const c = 200; // m/s
const n = 100;
const delta_x = L / n;
const delta_t = delta_x / (c * Math.sqrt(2)); // stability condition
const f = 1 / delta_t;
console.log('Sampling frequency is', f, 'Hz');

let y = pluck(0.1, 0.007);
let v = math.zeros(n).toArray();

// Prepare data for plotting
const x = [...Array(n).keys()].map(i => i * delta_x);

// Setup Plotly plot
const plotDiv = document.getElementById('plot');
const trace = {
    x: x,
    y: y,
    mode: 'lines',
    name: 'String Displacement',
    line: {color: 'red'}
};
const layout = {
    title: 'String Displacement Over Time',
    xaxis: {
        title: 'x (position along the string, in m)',
        range: [0, L]  // Set the range of the x-axis from 0 to L (length of the string)
    },
    yaxis: {
        title: 'y (vertical displacement, in m)',
        range: [-0.01, 0.01]  // Set the range of the y-axis from -0.01 to 0.01 meters
    },
    showlegend: false
};
Plotly.newPlot(plotDiv, [trace], layout);

// Update the plot in real-time
function updatePlot() {
    [y, v] = timestep(y, v);
    Plotly.animate(plotDiv, {
        data: [{y: y}],
        traces: [0],
        layout: {}
    }, {
        transition: {
            duration: 0,
        },
        frame: {
            duration: 0,
            redraw: true
        }
    });
}

// Run the simulation and update the plot every few milliseconds
setInterval(updatePlot, 1);

// Functions used in the simulation
function pluck(l, delta_y) {
    const y = math.zeros(n).toArray();
    const n_displaced = parseInt(n * (l / L));
    for (let i = 1; i < n_displaced; i++) {
        y[i] = (i / n_displaced) * delta_y;
    }
    for (let i = n_displaced; i < n - 1; i++) {
        y[i] = (1 - (i - n_displaced + 1) / (n - n_displaced)) * delta_y;
    }
    return y;
}

function partial_x2(y) {
    const y_right = y.slice(1);
    const y_left = y.slice(0, -1);
    y_right.push(0);
    y_left.unshift(0);
    return y_right.map((yr, i) => (yr - 2 * y[i] + y_left[i]) / Math.pow(delta_x, 2));
}

function timestep(y, v) {
    const d2y_dx2 = partial_x2(y);
    const d2y_dt2 = math.subtract(math.multiply(math.square(c), d2y_dx2), math.multiply(b, v));
    v = math.add(v, math.multiply(d2y_dt2, delta_t));
    y = math.add(y, math.multiply(v, delta_t));
    return [y, v];
}